samples: drivers: crypto: Add Support for TI MSPM0 AES module

Girinandha-M opened the pull request:
TI MSPM0 has AES accelerator for encryption and decryption

- Add support for crypto AES ECB and CBC mode of operation for encryption and 
decryption for TI MSPM0 AES module.


Signed-off-by: Girinandha Manivelpandiyan girinandha@linumiz.com

Girinandha-M committed:
dts: bindings: crypto: Add bindings for TI MSPM0 AES Module

Add bindings for TI MSPM0 AES Module.

Signed-off-by: Girinandha Manivelpandiyan <girinandha@linumiz.com>

Girinandha-M committed:
dts: arm: ti: mspm0g: Add AES node for TI MSPM0

Add devicetree node support for AES ( Advanced Encryption Standard )
node for TI MSPM0.

Signed-off-by: Girinandha Manivelpandiyan <girinandha@linumiz.com>

Girinandha-M committed:
samples: drivers: crypto: Add AES sample support for MSPM0 board.

Add board-specific overlay to enable the MSPM0 AES in the zephyr
crypto sample for lp_mspm0g3507 board.

Signed-off-by: Girinandha Manivelpandiyan <girinandha@linumiz.com>

tbursztyka commented on drivers/crypto/crypto_mspm0_aes.c:
```diff
@@ -0,0 +1,610 @@
+											\
+DEVICE_DT_INST_DEFINE(n, crypto_aes_init, NULL, &crypto_aes_data_##n,			\
+		&crypto_aes_config_##n, POST_KERNEL, CONFIG_CRYPTO_INIT_PRIORITY,	\
+		(void *)&crypto_enc_funcs);
```
all the lines below MSPM0_AES_DEFINE should  should be indented.

tbursztyka commented on drivers/crypto/crypto_mspm0_aes.c:
```diff
@@ -0,0 +1,610 @@
+
+	k_mutex_init(&data->device_mutex);
+
+	k_sem_init(&data->aes_done, 0, 1);
```
move these initialization to the instantiation macro

tbursztyka commented on include/zephyr/crypto/cipher.h:
See https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html 

It can take some time to get such changes in.
If you want to see your driver merged, you'll have to move away the 2 modes you added in it. (as well as this current commit)

Girinandha-M committed:
drivers: crypto: mspm0: Add support for TI MSPM0 AES module

add driver support for TI MSPM0 AES module

Introduced a hardware accelerated AES driver for the TI MSPM0
enable support for encryption and decryption using ECB and CBC mode
of operation.

Signed-off-by: Girinandha Manivelpandiyan <girinandha@linumiz.com>

jpanisbl commented on drivers/crypto/crypto_mspm0_aes.c:
```diff
@@ -0,0 +1,439 @@
+
+	ret = aes_hw_init(ctx);
+	if (ret != 0) {
+		goto cleanup;
```
I suggest adding some `LOG_ERR` here.
This comment also applies to other parts of the code where, currently, it might fail without understandable message.

d-philpot commented on dts/arm/ti/mspm0/g/mspm0g.dtsi:
```diff
@@ -38,6 +38,13 @@
 			};
 		};
 
+		aes: aes@40442000 {
```
Not all G-series device support AES. Please move this node into the corresponding family files where applicable

ssekar15 replied:
@d-philpot  AFAIK g150x, g310x, g350x and g51x all supports AES. Could you specify which doesn't has AES?

ssekar15 replied:
> Not all G-series device support AES. Please move this node into the corresponding family files where applicable

ACK

d-philpot commented on drivers/crypto/crypto_mspm0_aes.c:
```diff
@@ -0,0 +1,447 @@
+	const struct crypto_mspm0_aes_config *config = dev->config;
+
+	DL_AES_enablePower(config->regs);
+
```
Need delay of atleast 4 ULPCLK cycles following enabling power and before writing to other peripheral registers

parthitce replied:
`delay_cycles(CONFIG_MSPM0_PERIPH_STARTUP_DELAY);`

parthitce commented on drivers/crypto/crypto_mspm0_aes.c:
```diff
@@ -0,0 +1,461 @@
+		break;
+
+	default:
+		LOG_ERR("Mode Not Supported");
```
sanity to supported mode can be done in the entry of the function?

parthitce commented on drivers/crypto/crypto_mspm0_aes.c:
```diff
@@ -0,0 +1,461 @@
+	}
+
+	session->keylen = keylen;
+	session->in_use = true;
```
if session is the critical resource, the re-ordering and mark in_use early will have shorter critical locked path.

parthitce commented on drivers/crypto/crypto_mspm0_aes.c:
```diff
@@ -0,0 +1,461 @@
+		}
+
+		/* wait for AES operation completion */
+		ret = k_sem_take(&data->aes_done, AES_SEM_TIMEOUT);
```
we are here with mutex + sem and blocking all the sessions? Hardware cannot do multiple ops / sessions at the same time?

parthitce reviewed (approved):
LGTM

Girinandha-M commented:
@ssekar15 @valeriosetti ping for review. Thanks.
